### PR TITLE
Updated vgmm

### PR DESCRIPTION
**Pull request by mamikonyana**
This diff refactors variational finite and infinite mixture models, so they both use the same core methods. This is a step before Implementing a different faster variational algorithm for both of the classes.

### CODE REVIEW

**Comment by mamikonyana** [comment]
Adding Stochastic Variational Inference variants of Finite GMM, and DP GMM

Now we have total of 5 Gaussian Mixture Model Algorithms
- EM_GMM: finite model that uses Expectation Maximization algorithm to find the best fit
- MEAN_FIELD_GMM: finite model that uses Mean-Field Variational Inference
- MEAN_FIELD_DPGMM: inifinite model that uses Mean-Field Variational Inference
- SVI_GMM: finite model that uses Stochastic Variational Inference
-SVI_DPGMM: inifite model that uses Stochastic Variational Inference

I'm still not sure which one performs better. SVI algorithms are supposed to be faster, but they don't seem to work very well with small amounts of data. I'll still have to do experiments to figure out benefits and drawbacks. So for now I think we should have all of them.

**Comment by pbailis** [comment]
needs squash

File: src/main/java/macrobase/analysis/stats/distribution/MultivariateTDistribution.java [hunk 1]
```diff
@@ -16,7 +17,7 @@
     private int dimensions;
     private double multiplier;
 
-    public MultivariateTDistribution(RealVector mean, RealMatrix covarianceMatrix, int degreesOfFreedom) {
+    public MultivariateTDistribution(RealVector mean, RealMatrix covarianceMatrix, double degreesOfFreedom) {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/distribution/MultivariateTDistribution.java]: comment?

File: src/main/java/macrobase/analysis/stats/distribution/MultivariateTDistribution.java [hunk 2]
```diff
@@ -34,22 +35,14 @@ public MultivariateTDistribution(RealVector mean, RealMatrix covarianceMatrix, i
...
-        }
-        // using the that Gamma(n+1/2) / Gamma(n) ~= sqrt(n)
-        // http://mathworld.wolfram.com/images/equations/GammaFunction/NumberedEquation29.gif
-        if ((doubleNumerator - doubleDenominator) % 2 == 1) {
-            double j = 0.5 * doubleDenominator;
-            gammaRation *= Math.sqrt(j) *
-                    (1 - 1. / (8 * j) + 1. / (128 * j * j) + 5. / (1024 * j * j * j));
-        }
-        return gammaRation;
+    public static double halfGammaRatio(double doubleNumerator, double doubleDenominator) {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/distribution/MultivariateTDistribution.java]: can you provide a comment? googled, no immediate hits for a 'half gamma' distribution.

File: src/main/java/macrobase/analysis/stats/distribution/Wishart.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    private int D;  // dimensions
+
+    public Wishart(RealMatrix omega, double nu) {
+        this.omega = omega;
+        this.nu = nu;
+        this.D = omega.getColumnDimension();
+        this.logDetOmega = Math.log((new EigenDecomposition(omega)).getDeterminant());
+    }
+
+    public double lnB() {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/distribution/Wishart.java]: Again, comments would be good. No "B" in https://en.wikipedia.org/wiki/Wishart_distribution

File: src/main/java/macrobase/analysis/stats/distribution/Wishart.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package macrobase.analysis.stats.distribution;
+
+import org.apache.commons.math3.linear.EigenDecomposition;
+import org.apache.commons.math3.linear.RealMatrix;
+import org.apache.commons.math3.special.Gamma;
+
+public class Wishart {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/distribution/Wishart.java]: Is there any benefit in a common base `Distribution` class?

Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/distribution/Wishart.java]: Or Interface?

File: src/main/java/macrobase/analysis/stats/mixture/BatchMixtureModel.java [hunk 5]
```diff
@@ -68,12 +75,37 @@ public BatchMixtureModel(MacroBaseConf conf) {
...
      * @return covariances of mixture components
      */
     public abstract List<RealMatrix> getClusterCovariances();
 
     public abstract double[] getClusterProbabilities(Datum d);
+
+
+    @Override
+    public double getZScoreEquivalent(double zscore) {
+        throw new NotImplementedException("");
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/BatchMixtureModel.java]: Better exception message?

File: src/main/java/macrobase/analysis/stats/mixture/DPComponents.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
+package macrobase.analysis.stats.mixture;
+
+import org.apache.commons.math3.special.Gamma;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DPComponents implements MixingComponents {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/DPComponents.java]: Comment

File: src/main/java/macrobase/analysis/stats/mixture/DPGMM.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public void train(List<Datum> data) {
+        // 0. Initialize all approximating factors
+        log.debug("training locally");
+        clusters = new NormalWishartClusters(T, data.get(0).getMetrics().getDimension());
+        clusters.initializeBaseForDP(data);
+        clusters.initializeAtomsForDP(data, conf.getRandom());
+
+        log.debug("actual training");
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/DPGMM.java]: Nit: Necessary?

File: src/main/java/macrobase/analysis/stats/mixture/DPGMM.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        super(conf);
+        T = conf.getInt(MacroBaseConf.DPM_TRUNCATING_PARAMETER, MacroBaseDefaults.DPM_TRUNCATING_PARAMETER);
+        double concentrationParameter = conf.getDouble(MacroBaseConf.DPM_CONCENTRATION_PARAMETER, MacroBaseDefaults.DPM_CONCENTRATION_PARAMETER);
+        mixingComponents = new DPComponents(concentrationParameter, T);
+    }
+
+    @Override
+    public void train(List<Datum> data) {
+        // 0. Initialize all approximating factors
+        log.debug("training locally");
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/DPGMM.java]: Nit: actually necessary?

File: src/main/java/macrobase/analysis/stats/mixture/FiniteGMM.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Override
+    public void train(List<Datum> data) {
+        // 0. Initialize all approximating factors
+        log.debug("training locally");
+        mixingComponents = new MultiComponents(0.1, K);
+        clusters = new NormalWishartClusters(K, data.get(0).getMetrics().getDimension());
+        clusters.initializeBaseForFinite(data);
+        clusters.initalizeAtomsForFinite(data, initialClusterCentersFile, conf.getRandom());
+        //clusters.initializeBase(baseLoc, baseBeta, baseOmega, baseNu);
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/FiniteGMM.java]: dead, ditto log comments

File: src/main/java/macrobase/analysis/stats/mixture/NormalWishartClusters.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        loc = BatchMixtureModel.gonzalezInitializeMixtureCenters(data, K, random);
+        for (int i = 0; i < K; i++) {
+            // initialize betas as if all points are from the first cluster.
+            beta[i] = 1;
+            dof[i] = baseNu;
+            omega.add(0, AlgebraUtils.invertMatrix(baseOmegaInverse));
+        }
+    }
+
+    public void initalizeAtomsForFinite(List<Datum> data, String filename, Random random) {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/NormalWishartClusters.java]: comment

File: src/main/java/macrobase/analysis/stats/mixture/VariationalInference.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class VariationalInference {
+    private static final Logger log = LoggerFactory.getLogger(VariationalInference.class);
+
+    public static double[][] normalizeLogProbas(double[] lnMixing, double[] lnPrecision, double[][] dataLogLike) {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/VariationalInference.java]: probas?

**Comment by pbailis** [comment]
Generally LGTM, minor comments. 

Looking forward to syncing up this week to talk about what these actually do :)

File: src/main/java/macrobase/analysis/stats/mixture/BatchMixtureModel.java (hunk 5 above)
Comment (mamikonyana) [inline src/main/java/macrobase/analysis/stats/mixture/BatchMixtureModel.java]: this is a deprecated method, no? do you have a good suggestion?

Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/BatchMixtureModel.java]: We can remove it! For now, perhaps just a more descriptive message.

File: src/main/java/macrobase/analysis/stats/distribution/MultivariateTDistribution.java (hunk 2 above)
Comment (mamikonyana) [inline src/main/java/macrobase/analysis/stats/distribution/MultivariateTDistribution.java]: this just calculates `Gamma(num/2) / Gamma(den/2)`, I'll add a comment

File: src/main/java/macrobase/analysis/stats/distribution/Wishart.java (hunk 4 above)
Comment (mamikonyana) [inline src/main/java/macrobase/analysis/stats/distribution/Wishart.java]: yes, I will have an interface in the next PR.

File: src/main/java/macrobase/analysis/stats/distribution/MultivariateTDistribution.java (hunk 2 above)
Comment (mamikonyana) [inline src/main/java/macrobase/analysis/stats/distribution/MultivariateTDistribution.java]: I'll remove this.

**Pull request closed without merging**